extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerSketchFetcher.java

```diff
@@ -155,7 +152,17 @@ CompletableFuture<Either<Long, ClusterByPartitions>> inMemoryFullSketchMerging(
             mergedStatisticsCollector.clear();
           }
         }
-      });
+      }));
+    });
+
+    partitionFuture.whenComplete((result, exception) -> {
```

Comment:
in a corner case, it could happen that `partitionFuture` finishes before all futures are queued up in the executor. To avoid that one simple solution could be to have a `CountDownLatch` with value as `workerCount`. The `whenComplete` would then wait for latch to become 0 before cancelling all the futures.
Maybe we could also have a better solution with cancellation - will update on this thread if I find any utility.

Reply:
Is this a failsafe condition or would this be required? From my understanding, the successful completion of partitionFuture depends on the completion of the underlying futures, and in case of unsuccessful completion, we shouldn't have any issue with prematurely cancelling any underlying future. I might have missed something though.

Reply:
I also have a question in addition to Laksh's point, since we are attaching this callback in the same thread after queueing up the futures, would this trigger without all the futures present anyway?

Reply:
Yes, agree that since attaching this callback is in the same thread after queuing - this shouldn't trigger without futures. Although, that raises the question that if the future finishes before this callback is attached, will this callback execute while attaching it? if so, will that callback execution happen on the same thread attaching it?

Reply:
Also, is it necessary to cancel the pending tasks? It would be incase we'll be retrying the stats collection stage - otherwise if we're doing to fail the whole job, it is not necessary to fail them

Reply:
Yes, the future that is added as a callback should still get executed if the completable future is already finished executing, though the javadoc doesn't guarantee which thread would be executing it if an executor is also not passed as an argument.

It is also not a hard necessity that we cancel every pending future but it depends on number of workers which could be up to a 1000, which would waste a lot of time, but correctness would still be present.

 Also, with regard to retrying, part of this will be updated with the worker retry PR, to only fetch the statistics which are needed even if a worker restarts, so after that we would be retrying the stats collection.